File: services/account.go
package services

import (
	"bank_api_go/apperrors"
	"bank_api_go/entities"

	"gorm.io/gorm"
)

type AccountService struct {
	db          *gorm.DB
	idempotency *IdempotencyService
}

func NewAccountService(db *gorm.DB, idempotency *IdempotencyService) *AccountService {
	return &AccountService{db: db, idempotency: idempotency}
}

func (s *AccountService) FindByID(id uint) (*entities.Account, error) {
	var account entities.Account
	if err := s.db.First(&account, id).Error; err != nil {
		return nil, &apperrors.AccountNotFound{Message: "Account not found"}
	}
	return &account, nil
}

func (s *AccountService) Create(idempotencyKey string, fingerprint string, account *entities.Account) (*entities.Account, error) {
	err := s.db.Transaction(func(tx *gorm.DB) error {
		if err := s.idempotency.EnsureUnique(tx, idempotencyKey, fingerprint); err != nil {
			return err
		}

		if err := tx.Create(account).Error; err != nil {
			return err
		}

		return s.idempotency.Save(tx, idempotencyKey, fingerprint, 201, account)
	})
	if err != nil {
		return nil, err
	}

	return account, nil
}

func (s *AccountService) Deposit(idempotencyKey string, fingerprint string, id uint, amount int64) (*entities.Account, error) {
	var account entities.Account
	err := s.db.Transaction(func(tx *gorm.DB) error {
		if err := s.idempotency.EnsureUnique(tx, idempotencyKey, fingerprint); err != nil {
			return err
		}

		if err := tx.First(&account, id).Error; err != nil {
			return &apperrors.AccountNotFound{Message: "Account not found"}
		}

		if err := tx.Model(&account).Update("balance", gorm.Expr("balance + ?", amount)).Error; err != nil {
			return err
		}
		account.Balance += amount

		return s.idempotency.Save(tx, idempotencyKey, fingerprint, 200, &account)
	})
	if err != nil {
		return nil, err
	}

	return &account, nil
}

func (s *AccountService) Transfer(idempotencyKey string, fingerprint string, fromID, toID uint, amount int64) (*entities.TransferResult, error) {
	if fromID == toID {
		return nil, &apperrors.BadRequest{Message: "Cannot transfer to the same account"}
	}

	var result entities.TransferResult
	err := s.db.Transaction(func(tx *gorm.DB) error {
		if err := s.idempotency.EnsureUnique(tx, idempotencyKey, fingerprint); err != nil {
			return err
		}

		if err := tx.First(&result.From, fromID).Error; err != nil {
			return &apperrors.AccountNotFound{Message: "Source account not found"}
		}
		if err := tx.First(&result.To, toID).Error; err != nil {
			return &apperrors.AccountNotFound{Message: "Destination account not found"}
		}

		if result.From.Currency != result.To.Currency {
			return &apperrors.BadRequest{Message: "Cannot transfer between accounts with different currencies"}
		}

		if result.From.Balance < amount {
			return &apperrors.InsufficientFunds{}
		}

		if err := tx.Model(&result.From).Update("balance", gorm.Expr("balance - ?", amount)).Error; err != nil {
			return err
		}
		result.From.Balance -= amount

		if err := tx.Model(&result.To).Update("balance", gorm.Expr("balance + ?", amount)).Error; err != nil {
			return err
		}
		result.To.Balance += amount

		return s.idempotency.Save(tx, idempotencyKey, fingerprint, 200, &result)
	})
	if err != nil {
		return nil, err
	}

	return &result, nil
}

File: ts_solution/src/services/account.ts
import { AppDataSource } from "../db/data-source";
import { Account } from "../entities/Account";
import { IdempotencyRecord } from "../entities/IdempotencyRecord";
import NotFound from "../errors/NotFound";
import InsufficientFunds from "../errors/InsufficientFunds";
import BadRequest from "../errors/BadRequest";
import DuplicateRequest from "../errors/DuplicateRequest";

class AccountService {
    repo: any;

    constructor() {
        this.repo = AppDataSource.getRepository(Account);
    }

    async findById(id: number): Promise<Account> {
        const account = await this.repo.findOneBy({ id: Number(id) });
        if(!account) {
            throw new NotFound();
        }
        return account;
    }

    async create(idempotencyKey: string, payload: Partial<Account>): Promise<Account> {
        await this.ensureIdempotency(idempotencyKey);
        const account = this.repo.create(payload);
        const saved = await this.repo.save(account);
        await this.saveIdempotency(idempotencyKey, 201, saved);
        return saved;
    }

    async deposit(idempotencyKey: string, id: number, amount: number): Promise<Account> {
        await this.ensureIdempotency(idempotencyKey);

        const saved = await AppDataSource.transaction(async (manager) => {
            const account = await manager.findOneBy(Account, { id });
            if (!account) throw new NotFound();

            account.balance += amount;
            return manager.save(account);
        });

        await this.saveIdempotency(idempotencyKey, 200, saved);
        return saved;
    }

    async transfer(idempotencyKey: string, fromId: number, toId: number, amount: number): Promise<{ from: Account; to: Account }> {
        if (fromId === toId) {
            throw new BadRequest("Cannot transfer to the same account");
        }

        await this.ensureIdempotency(idempotencyKey);

        const result = await AppDataSource.transaction(async (manager) => {
            const from = await manager.findOneBy(Account, { id: fromId });
            if (!from) throw new NotFound("Source account not found");

            const to = await manager.findOneBy(Account, { id: toId });
            if (!to) throw new NotFound("Destination account not found");

            if (from.balance < amount) {
                throw new InsufficientFunds();
            }

            from.balance -= amount;
            to.balance += amount;

            await manager.save(from);
            await manager.save(to);

            return { from, to };
        });

        await this.saveIdempotency(idempotencyKey, 200, result);
        return result;
    }

    private async ensureIdempotency(key: string): Promise<void> {
        const record = await AppDataSource.getRepository(IdempotencyRecord).findOneBy({ key });
        if (record) {
            throw new DuplicateRequest(record.statusCode, JSON.parse(record.response));
        }
    }

    private async saveIdempotency(key: string, statusCode: number, response: any): Promise<void> {
        const record = new IdempotencyRecord();
        record.key = key;
        record.statusCode = statusCode;
        record.response = JSON.stringify(response);
        await AppDataSource.getRepository(IdempotencyRecord).save(record);
    }
}

const accountService = new AccountService();

export default accountService;
